Orchestrator.py

```diff
@@ -747,4 +783,151 @@ def _computeHeadingSpeedMovementTimeout(self, dotBotId, path):
             movementTimeout)
         )
 
-        return (heading, speed, movementTimeout)
\ No newline at end of file
+        return (heading, speed, movementTimeout)
+
+    # === Relays
+
+    def _assignRelaysAndRelayPositionsCb(self):
+
+        log.debug('heartbeats {}'.format([db['heartbeat'] for (_, db) in self.dotBotsView.items()]))
+        self.simEngine.schedule(self.simEngine.currentTime() + 10, self._assignRelaysAndRelayPositionsCb)
+
+        if self.relayAlgorithm   == "Recovery":
+            self._recoveryRelayPlacement()
+
+        elif self.relayAlgorithm == "SelfHealing":
+            self._selfHealingRelayPlacement()
+
+        else:
+            # don't place relays if no algorithm is specified
+            pass
+
+    def _recoveryRelayPlacement(self):
+
+        # first check if we need relays
+        for (dotBotId, dotBot) in self.dotBotsView.items():
+
+            # if all DotBot PDRs are above min threshold don't assign a relay
```

Comment:
comment is wrong, what you code does it is: "DotBot with high PDR isn't a relay"